File: contractsdk/go/code/context.go
package code

import (
	"math/big"

	pb "github.com/xuperchain/xuperunion/contractsdk/go/pb"
)

// Context is the context in which the contract runs
type Context interface {
	Args() map[string][]byte
	Caller() string
	Initiator() string
	AuthRequire() []string

	PutObject(key []byte, value []byte) error
	GetObject(key []byte) ([]byte, error)
	DeleteObject(key []byte) error
	NewIterator(start, limit []byte) Iterator

	QueryTx(txid string) (*pb.TxStatus, error)
	QueryBlock(blockid string) (*pb.Block, error)
	Transfer(to string, amount *big.Int) error
	Call(module, contract, method string, args map[string][]byte) (*Response, error)
}

// Iterator iterates over key/value pairs in key order
type Iterator interface {
	Key() []byte
	Value() []byte
	Next() bool
	Error() error
	// Iterator 必须在使用完毕后关闭
	Close()
}

// PrefixRange returns key range that satisfy the given prefix
func PrefixRange(prefix []byte) ([]byte, []byte) {
	var limit []byte
	for i := len(prefix) - 1; i >= 0; i-- {
		c := prefix[i]
		if c < 0xff {
			limit = make([]byte, i+1)
			copy(limit, prefix)
			limit[i] = c + 1
			break
		}
	}
	return prefix, limit
}

File: core/contractsdk/java/src/main/java/com/baidu/xuper/Context.java
package com.baidu.xuper;

import com.baidu.xuper.contractpb.Contract;

import java.math.BigInteger;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public interface Context {
    public Map<String, byte[]> args();

    public String caller();

    public List<String> authRequire();

    public void putObject(byte[] key, byte[] value);

    public byte[] getObject(byte[] key);

    public void deleteObject(byte[] key);

    public Iterator<ContractIteratorItem> newIterator(byte[] start, byte[] limit);

    public Contract.Transaction queryTx(String txid);

    public Contract.Block queryBlock(String blockid);

    public void transfer(String to, BigInteger amount);

    public BigInteger transferAmount();

    public Response call(String module, String contract, String method, Map<String, byte[]> args);

    public Response crossQuery(String uri, Map<String, byte[]> args);

    public void emitEvent(String name, byte[] body);

    public void emitJSONEvent(String name, Map<String, byte[]> body);

    public void log(String msg);
}